File: java/src/n04_observer/Main.java
package n04_observer;

import java.io.File;
import java.net.URISyntaxException;

public class Main {
    public static void main(String[] args) throws URISyntaxException {
        Editor editor = new Editor();
        editor.events.subscribe(EventType.OPEN, new LoggingListener());
        editor.events.subscribe(EventType.SAVE, new LoggingListener());
        editor.events.subscribe(EventType.SAVE, new EMailAlertsListener());

        File file = new File(Main.class.getProtectionDomain().getCodeSource().getLocation().toURI());
        editor.openFile(file);
        editor.saveFile(file);
    }
}

File: dart/src/n04_observer/main.dart
import 'dart:io';

import 'editor.dart';
import 'email_alerts_listener.dart';
import 'event_type.dart';
import 'logging_listener.dart';

void main() {
    final Editor editor = Editor();
    editor.events
        ..subscribe(EventType.open, LoggingListener())
        ..subscribe(EventType.save, LoggingListener())
        ..subscribe(EventType.save, EMailAlertsListener());

    final File file = File(Platform.script.toFilePath());
    editor
        ..openFile(file)
        ..saveFile(file);
}
